azurelinuxagent/common/protocol/extensions_goal_state_from_vm_settings.py

```diff
@@ -325,7 +335,52 @@ def _parse_extensions(self, vm_settings):
                         raise Exception('additionalLocations should be an array')
                     extension.manifest_uris.extend(additional_locations)
 
-                # TODO: Add settings
+                #
+                # Settings
+                #
+                settings_list = extension_gs.get('settings')
+                if settings_list is not None:
+                    if not isinstance(settings_list, list):
+                        raise Exception("'settings' should be an array (extension: {0})".format(extension.name))
+                    if not extension.supports_multi_config and len(settings_list) > 1:
+                        raise Exception("Single-config extension includes multiple settings (extension: {0})".format(extension.name))
+
+                    for s in settings_list:
+                        settings = ExtensionSettings()
+                        public_settings = s.get('publicSettings')
+                        settings.publicSettings = {} if public_settings is None else json.loads(public_settings)
+                        # Note that protectedSettings and protectedSettingsCertThumbprint can be None (which would be serialized to the extension's settings file as null)
+                        settings.protectedSettings = s.get('protectedSettings')
+                        thumbprint = s.get('protectedSettingsCertThumbprint')
+                        if thumbprint is None and settings.protectedSettings is not None:
```

Comment:
we should probably check for the opposite case too, i.e. - 

`if thumbprint is not None and settings.protectedSettings is None`
```suggestion
                        (if thumbprint is not None and settings.protectedSettings is None) or (if thumbprint is None and settings.protectedSettings is not None):
```

Reply:
protected settings without a cert would definitively be an error since there would be no way to decrypt the settings

not sure if CRP would ever send a thumbprint without protected settings, so i decided to be cautious and did not add that check

Reply:
(btw the current parsing of extensions config does not do any checks and simply sets the properties to None)

Reply:
> not sure if CRP would ever send a thumbprint without protected settings, so i decided to be cautious and did not add that check

Fair enough, nothing would go wrong either I suppose if we have a cert but no one asking for it haha